File: android/src/main/java/com/supermap/rnsupermap/JSGeoStyle.java
package com.supermap.rnsupermap;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.supermap.data.Color;
import com.supermap.data.GeoStyle;
import com.supermap.data.Size2D;

import java.util.Calendar;
import java.util.HashMap;
import java.util.Map;

public class JSGeoStyle extends ReactContextBaseJavaModule {
    public static final String REACT_CLASS = "JSGeoStyle";
    private static Map<String, GeoStyle> m_GeoStyleList = new HashMap<String, GeoStyle>();
    GeoStyle m_GeoStyle;

    public JSGeoStyle(ReactApplicationContext context) {
        super(context);
    }

    public static GeoStyle getObjFromList(String id) {
        return m_GeoStyleList.get(id);
    }

    @Override
    public String getName() {
        return REACT_CLASS;
    }

    public static String registerId(GeoStyle obj) {
        for (Map.Entry entry : m_GeoStyleList.entrySet()) {
            if (obj.equals(entry.getValue())) {
                return (String) entry.getKey();
            }
        }

        Calendar calendar = Calendar.getInstance();
        String id = Long.toString(calendar.getTimeInMillis());
        m_GeoStyleList.put(id, obj);
        return id;
    }

    @ReactMethod
    public void createObj(Promise promise){
        try{
            GeoStyle geoStyle = new GeoStyle();
            String geoStyleId = registerId(geoStyle);

            WritableMap map = Arguments.createMap();
            map.putString("geoStyleId",geoStyleId);
            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void setLineColor(String geoStyleId,int r,int g,int b,Promise promise){
        try{
            GeoStyle geoStyle = m_GeoStyleList.get(geoStyleId);
            Color color = new Color(r,g,b);
            geoStyle.setLineColor(color);

            promise.resolve(true);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void setLineSymbolID(String geoStyleId,int symboleId,Promise promise){
        try{
            GeoStyle geoStyle = m_GeoStyleList.get(geoStyleId);
            geoStyle.setLineSymbolID(symboleId);

            promise.resolve(true);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void setLineWidth(String geoStyleId,Double lineWidth,Promise promise){
        try{
            GeoStyle geoStyle = m_GeoStyleList.get(geoStyleId);
            geoStyle.setLineWidth(lineWidth);

            promise.resolve(true);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void setMarkerSymbolID(String geoStyleId,int markerSymbolId,Promise promise){
        try{
            GeoStyle geoStyle = m_GeoStyleList.get(geoStyleId);
            geoStyle.setMarkerSymbolID(markerSymbolId);

            promise.resolve(true);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void setMarkerSize(String geoStyleId,String size2DId,Promise promise){
        try{
            GeoStyle geoStyle = m_GeoStyleList.get(geoStyleId);
            Size2D size2D = JSSize2D.getObjFromList(size2DId);
            geoStyle.setMarkerSize(size2D);

            promise.resolve(true);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void setFillForeColor(String geoStyleId,int r,int g,int b,Promise promise){
        try{
            GeoStyle geoStyle = m_GeoStyleList.get(geoStyleId);
            geoStyle.setFillForeColor(new Color(r,g,b));

            promise.resolve(true);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void setFillOpaqueRate(String geoStyleId,int rate,Promise promise){
        try{
            GeoStyle geoStyle = m_GeoStyleList.get(geoStyleId);
            geoStyle.setFillOpaqueRate(rate);

            promise.resolve(true);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void getLineColor(String geoStyleId, Promise promise){
        try{
            GeoStyle geoStyle = m_GeoStyleList.get(geoStyleId);
            Color color = geoStyle.getLineColor();
            WritableMap map = Arguments.createMap();
            map.putInt("r", color.getR());
            map.putInt("g", color.getG());
            map.putInt("b", color.getB());
            map.putInt("a", color.getA());

            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void getLineSymbolID(String geoStyleId, Promise promise){
        try{
            GeoStyle geoStyle = m_GeoStyleList.get(geoStyleId);
            int value = geoStyle.getLineSymbolID();

            promise.resolve(value);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void getLineWidth(String geoStyleId, Promise promise){
        try{
            GeoStyle geoStyle = m_GeoStyleList.get(geoStyleId);
            double value = geoStyle.getLineWidth();

            promise.resolve(value);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void getMarkerSymbolID(String geoStyleId, Promise promise){
        try{
            GeoStyle geoStyle = m_GeoStyleList.get(geoStyleId);
            int value = geoStyle.getMarkerSymbolID();

            promise.resolve(value);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void getMarkerSize(String geoStyleId, Promise promise){
        try{
            GeoStyle geoStyle = m_GeoStyleList.get(geoStyleId);
            Size2D size2D = geoStyle.getMarkerSize();
            String size2DId = JSSize2D.registerId(size2D);

            promise.resolve(size2DId);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void getFillForeColor(String geoStyleId, Promise promise){
        try{
            GeoStyle geoStyle = m_GeoStyleList.get(geoStyleId);
            Color color = geoStyle.getFillForeColor();
            WritableMap map = Arguments.createMap();
            map.putInt("r", color.getR());
            map.putInt("g", color.getG());
            map.putInt("b", color.getB());
            map.putInt("a", color.getA());

            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void getFillOpaqueRate(String geoStyleId, Promise promise){
        try{
            GeoStyle geoStyle = m_GeoStyleList.get(geoStyleId);
            int value = geoStyle.getFillOpaqueRate();

            promise.resolve(value);
        }catch (Exception e){
            promise.reject(e);
        }
    }
}

File: ios/objNative/JSGeoStyle.m
//
//  JSGeoStyle.m
//  HelloWorldDemo
//
//  Created by 王子豪 on 2016/11/22.
//  Copyright © 2016年 Facebook. All rights reserved.
//

#import "JSGeoStyle.h"
#import "SuperMap/GeoStyle.h"
#import "SuperMap/Color.h"
#import "JSObjManager.h"
@implementation JSGeoStyle
RCT_EXPORT_MODULE();
RCT_REMAP_METHOD(createObj,createObjWithresolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
  GeoStyle* style = [[GeoStyle alloc]init];
  if(style){
    NSInteger key = (NSInteger)style;
    [JSObjManager addObj:style];
    resolve(@{@"geoStyleId":@(key).stringValue});
  }else{
    reject(@"geoStyle",@"create geoStyle failed!!!",nil);
  }
}

RCT_REMAP_METHOD(setLineColor,setLineColorWithStyleId:(NSString*)styleId redColor:(int)red greenColor:(int)green blueColor:(int)blue resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    Color* color = [[Color alloc]initWithR:red G:green B:blue];
    GeoStyle* style = [JSObjManager getObjWithKey:styleId];
  if (style) {
    [style setLineColor:color];
    resolve(@"1");
  }else{
     reject(@"geoStyle",@"setLineColor failed!!!",nil);
  }
}
RCT_REMAP_METHOD(setPointColor,setPointColorId:(NSString*)styleId redColor:(int)red greenColor:(int)green blueColor:(int)blue resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    Color* color = [[Color alloc]initWithR:red G:green B:blue];
    GeoStyle* style = [JSObjManager getObjWithKey:styleId];
    if (style) {
        [style setLineColor:color];
       // [style setPointColor:color];
        resolve(@"1");
    }else{
        reject(@"geoStyle",@"setLineColor failed!!!",nil);
    }
}
RCT_REMAP_METHOD(setLineSymbolID,setLineSymbolIdWithStyleId:(NSString*)styleId symbolId:(int)symbolId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    GeoStyle* style = [JSObjManager getObjWithKey:styleId];
    if (style) {
        [style setLineSymbolID:symbolId];
        resolve(@"1");
    }else{
        reject(@"geoStyle",@"setLineSymbolID failed!!!",nil);
    }
}


RCT_REMAP_METHOD(getPointColor,getPointColorId:(NSString*)styleId  resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
  GeoStyle* style = [JSObjManager getObjWithKey:styleId];
  if (style) {
      Color* color = [style getLineColor];
      resolve(@{@"r":@(color.red),
                @"g":@(color.green),
                @"b":@(color.blue),
                });
  }else{
    reject(@"geoStyle",@"setLineSymbolID failed!!!",nil);
  }
}

RCT_REMAP_METHOD(setLineWidth,setLineWidthWithStyleId:(NSString*)styleId lineWidth:(double)lineWidth resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
  GeoStyle* style = [JSObjManager getObjWithKey:styleId];
  if (style) {
    [style setLineWidth:lineWidth];
    resolve(@"1");
  }else{
    reject(@"geoStyle",@"setLineWidth failed!!!",nil);
  }
}

RCT_REMAP_METHOD(setMarkerSymbolID,setMarkerSymbolIdWithStyleId:(NSString*)styleId markerSymbolId:(int)markerSymbolId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        GeoStyle* style = [JSObjManager getObjWithKey:styleId];
        [style setMarkerSymbolID:markerSymbolId];
        resolve([NSNumber numberWithBool:YES]);
    } @catch (NSException* exception) {
        reject(@"geoStyle",@"setMarkerSymbolID failed!!!",nil);
    }
}
////////////////
RCT_REMAP_METHOD(setMarkerSize,setMarkerSizeWithStyleId:(NSString*)styleId sizeId:(NSString*)sizeId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
  GeoStyle* style = [JSObjManager getObjWithKey:styleId];
  Size2D* size = [JSObjManager getObjWithKey:sizeId];
  if (style) {
    [style setMarkerSize:size];
    resolve(@"1");
  }else{
    reject(@"geoStyle",@"setMarkerSize failed!!!",nil);
  }
}

RCT_REMAP_METHOD(setFillForeColor,setFillForeColorWithStyleId:(NSString*)styleId redColor:(int)redColor greenColor:(int)greenColor blueColor:(int)blueColor resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
  GeoStyle* style = [JSObjManager getObjWithKey:styleId];
  Color* color = [[Color alloc]initWithR:redColor G:greenColor B:blueColor];
  if (style) {
    [style setFillForeColor:color];
    resolve(@"1");
  }else{
    reject(@"geoStyle",@"setFillForeColor failed!!!",nil);
  }
}

RCT_REMAP_METHOD(setFillOpaqueRate,setFillOpaqueRateWithStyleId:(NSString*)styleId rate:(int)rate resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
  GeoStyle* style = [JSObjManager getObjWithKey:styleId];
  if (style) {
    [style setFillOpaqueRate:rate];
    resolve(@"1");
  }else{
    reject(@"geoStyle",@"setFillForeColor failed!!!",nil);
  }
}

RCT_REMAP_METHOD(getLineColor,getLineColorId:(NSString*)styleId  resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    GeoStyle* style = [JSObjManager getObjWithKey:styleId];
    if (style) {
        Color* color = [style getLineColor];
        resolve(@{@"r":@(color.red),
                  @"g":@(color.green),
                  @"b":@(color.blue),
                  @"a":@(color.alpha)
                  });
    }else{
        reject(@"geoStyle",@"jsGeoStyle getLineColor failed!!!",nil);
    }
}
RCT_REMAP_METHOD(getLineSymbolID,getLineSymbolIDId:(NSString*)styleId  resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    GeoStyle* style = [JSObjManager getObjWithKey:styleId];
    if (style) {
        int color = [style getLineSymbolID];
        resolve(@(color));
    }else{
        reject(@"geoStyle",@"jsGeoStyle getLineSymbolID failed!!!",nil);
    }
}

RCT_REMAP_METHOD(getLineWidth,getLineWidthId:(NSString*)styleId  resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    GeoStyle* style = [JSObjManager getObjWithKey:styleId];
    if (style) {
        double color = [style getLineWidth];
        resolve(@(color));
    }else{
        reject(@"geoStyle",@"jsGeoStyle getLineWidth failed!!!",nil);
    }
}

RCT_REMAP_METHOD(getMarkerSymbolID,getMarkerSymbolIDId:(NSString*)styleId  resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    GeoStyle* style = [JSObjManager getObjWithKey:styleId];
    if (style) {
        int color = [style getMarkerSymbolID];
        resolve(@(color));
    }else{
        reject(@"geoStyle",@"jsGeoStyle getLineWidth failed!!!",nil);
    }
}

RCT_REMAP_METHOD(getMarkerSize,getMarkerSizeId:(NSString*)styleId  resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    GeoStyle* style = [JSObjManager getObjWithKey:styleId];
    if (style) {
        Size2D* color = [style getMarkerSize];
        resolve([JSObjManager addObj:color]);
    }else{
        reject(@"geoStyle",@"jsGeoStyle getMarkerSize failed!!!",nil);
    }
}

RCT_REMAP_METHOD(getFillForeColor,getFillForeColorId:(NSString*)styleId  resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    GeoStyle* style = [JSObjManager getObjWithKey:styleId];
    if (style) {
        Color* color = [style getFillForeColor];
        resolve(@{@"r":@(color.red),
                  @"g":@(color.green),
                  @"b":@(color.blue),
                  @"a":@(color.alpha)
                  });

    }else{
        reject(@"geoStyle",@"jsGeoStyle getFillForeColor failed!!!",nil);
    }
}

RCT_REMAP_METHOD(getFillOpaqueRate,getFillOpaqueRateId:(NSString*)styleId  resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    GeoStyle* style = [JSObjManager getObjWithKey:styleId];
    if (style) {
        double color = [style getFillOpaqueRate];
        resolve(@(color));

    }else{
        reject(@"geoStyle",@"jsGeoStyle getMarkerSize failed!!!",nil);
    }
}

@end
